### DISCUSSION TITLE
stat or numstat

### DISCUSSION THREAD

**Discussion by ghost** [jelmer/dulwich#944] [Q&A] [answered]
I am looking for commit stats. like provided by `git log` or `git diff`. I can not find anything not even a example how to do that myself using dulwich.

My end goal is to get the line numbers per changed file for all modifications (add, delete, ...). Any tips would be appreciated!

**Comment by jelmer** [comment, marked as answer]
There are ``porcelain.log()`` and ``porcelain.diff`` that provide behaviour similar to "git log" and "git diff".

Are you looking for the number of deleted/added lines per file for every comnmit? If so, you could probably iterate over all relevant commits with ``Repo.walk``, then generate a diff using the functions in ``dulwich.patch` and then filter the whole patch through something like the diffstat utility (or a homegrown implementation that looks for leading + and - signs) to get the number of changed lines. Let me know if I can provide more specific pointers.

**Comment by ghost** [comment]
thank you! it works now. I did like you suggested and used slightly modified functions from `dulwich.patch` (mostly unified_diff). I now get the statistics I want (I needed `the "magnitude" of changes in each file and also the exact position of the change` so I can attribute the change to a function, a class etc.). Cool thing is that with dulwich I do not have to use three different technologies (bash, git, python, pandas) to achieve this.

**Comment by jelmer** [comment]
Thanks, great to hear that worked. :)
